packages/babylon/src/index.js

```diff
@@ -18,7 +18,27 @@ plugins.jsx = jsxPlugin;
 plugins.typescript = typescriptPlugin;
 
 export function parse(input: string, options?: Options): File {
-  return getParser(options, input).parse();
+  if (options && options.sourceType === "unambiguous") {
+    options = Object.assign({}, options);
+    try {
+      options.sourceType = "module";
+      const ast = getParser(options, input).parse();
+
+      // Rather than try to parse as a script first, we opt to parse as a module and convert back
+      // to a script where possible to avoid having to do a full re-parse of the input content.
+      if (!hasModuleSyntax(ast)) ast.program.sourceType = "script";
+      return ast;
+    } catch (moduleError) {
+      try {
+        options.sourceType = "script";
+        return getParser(options, input).parse();
+      } catch (scriptError) {}
+
+      throw moduleError;
+    }
```

Comment:
👆 This aligns with [my approach](https://github.com/standard-things/esm/blob/bf531b69a408111fa6e66ac033a2230394483392/src/compiler.js#L68-L91) as well.